torch/distributed/_functional_collectives.py

```diff
@@ -363,26 +363,52 @@ def reduce_scatter_tensor(
         tensor_list = torch.chunk(self, group_size, dim=scatter_dim)
         self = torch.cat(tensor_list)
 
-    tensor = torch._C._nn.reduce_scatter_tensor(self, reduceOp, 0, tag, rankset, group_size)  # type: ignore[attr-defined]
+    tensor = torch.ops.c10d_functional.reduce_scatter_tensor(self, reduceOp, 0, tag, rankset, group_size)  # type: ignore[attr-defined]
     res = _maybe_wrap_tensor(tensor)
     return res
 
-
-c10_lib_cpu = torch.library.Library("aten", "IMPL", "CPU")
-c10_lib_cuda = torch.library.Library("aten", "IMPL", "CUDA")
+c10_lib = torch.library.Library("c10d_functional", "DEF")
+c10_lib_impl = torch.library.Library("c10d_functional", "IMPL")
 
 def _register_ops():
-    c10_lib_cpu.impl("all_reduce", _all_reduce)
-    c10_lib_cuda.impl("all_reduce", _all_reduce)
+    c10_lib.define("all_reduce(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor")
+    c10_lib.define("wait_tensor(Tensor self) -> Tensor")
+    c10_lib.define("all_gather_into_tensor(Tensor shard, str tag, int[] ranks, int group_size) -> Tensor")
+    c10_lib.define(
+        "reduce_scatter_tensor(Tensor input, str reduceOp, int scatter_dim, str tag, int[] ranks, int group_size) -> Tensor"
+    )
+
+    c10_lib_impl.impl("all_reduce", _all_reduce, "CPU")
```

Comment:
did we ever try using BackendSelect here?  Would that even make better sense?  In the limit i think its ok to assume we trace this on any backend and then if the backend doesn't support allreduce it errs.  (Compared to having to require any additional backend to also register here)

Reply:
I just tried and it works. Switching over it.